minio/minio
# Enable replication of SSE-C objects

## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license](https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
If site replication enabled across sites, replicate the SSE-C objects as well. These objects could be read from target sites using the same client encryption keys.

## Motivation and Context
Support replication of client encrypted object to replicated sites.

## How to test this PR?
Needs https://github.com/minio/minio-go/pull/1943

1. Set up two MinIO sites with TLS enabled

2. Create a huge (~500MiB) file with known content
```
for index in {1..10000000}; do echo "$index - The quick brown fox jumps over the lazy dog" >> hugefile;         done
```

3. Create a bucket in first site and add objects, two with client side encryption and one without any encryption
```
$ mc cp /etc/hosts ALIAS1/BUCKET --insecure
$ mc cp /etc/issue ALIAS1/BUCKET --encrypt-key "ALIAS1/BUCKET/issue=iliketobecrazybutnotsomuchreally" --insecure`
$ mc cp ./hugefile ALIAS1/BUCKET --encrypt-key "ALIAS1/BUCKET/hugefile=iliketobecrazybutnotsomuchreally" --insecure
```

4. Setup site replication between the two sites
`mc admin replicate add ALIAS1 ALIAS2 --insecure`

5. List and see if bothe objects are present on both the sites
```
$ mc ls ALIAS1/BUCKET --insecure
[2024-02-28 12:19:43 IST] 1.0KiB STANDARD hosts
[2024-02-28 12:19:38 IST] 514MiB STANDARD hugefile
[2024-02-28 12:19:35 IST]    28B STANDARD issue

$ mc ls ALIAS2/BUCKET --insecure 
[2024-02-28 12:19:43 IST] 1.0KiB STANDARD hosts
[2024-02-28 12:19:38 IST] 514MiB STANDARD hugefile
[2024-02-28 12:19:35 IST]    28B STANDARD issue
```

6. Stat client encrypted objects from both the sites and make sure it works as expected
```
$ mc stat ALIAS1/BUCKET/issue --encrypt-key "ALIAS1/BUCKET/issue=iliketobecrazybutnotsomuchreally" --insecure
Name      : issue
Date      : 2024-02-28 12:19:35 IST 
Size      : 28 B   
ETag      : 6b47c0b518de13c30efcc7dc3111d57a 
Type      : file 
Encryption: SSE-C
Metadata  :
  Content-Type: application/octet-stream 
Replication Status: COMPLETED

$ mc stat ALIAS1/BUCKET/hugefile --encrypt-key "ALIAS1/BUCKET/hugefile=iliketobecrazybutnotsomuchreally" --insecure
Name      : hugefile
Date      : 2024-02-28 12:19:38 IST 
Size      : 514 MiB 
ETag      : 2476c8a936a3732a8b046d3376a5de9c-33 
Type      : file 
Encryption: SSE-C
Metadata  :
  Content-Type: application/octet-stream 
Replication Status: COMPLETED  

$ mc stat ALIAS2/BUCKET/issue --encrypt-key "ALIAS2/BUCKET/issue=iliketobecrazybutnotsomuchreally" --insecure
Name      : issue
Date      : 2024-02-28 12:19:35 IST 
Size      : 28 B   
ETag      : 6b47c0b518de13c30efcc7dc3111d57a 
Type      : file 
Encryption: SSE-C
Metadata  :
  Content-Type: application/octet-stream 
Replication Status: REPLICA 

$ mc stat ALIAS2/BUCKET/hugefile --encrypt-key "ALIAS2/BUCKET/hugefile=iliketobecrazybutnotsomuchreally" --insecure
Name      : hugefile
Date      : 2024-02-28 12:19:38 IST 
Size      : 514 MiB 
ETag      : 2476c8a936a3732a8b046d3376a5de9c-33 
Type      : file 
Encryption: SSE-C
Metadata  :
  Content-Type: application/octet-stream 
Replication Status: REPLICA
```

7. Cat the client encrypted objects from both sites and it should work as expected
```
$ mc cat ALIAS1/BUCKET/issue --encrypt-key "ALIAS1/BUCKET/issue=iliketobecrazybutnotsomuchreally" --insecure
\S
Kernel \r on an \m (\l)

$ mc cat ALIAS2/BUCKET/issue --encrypt-key "ALIAS2/BUCKET/issue=iliketobecrazybutnotsomuchreally" --insecure
\S
Kernel \r on an \m (\l)

$ mc cat ALIAS1/BUCKET/hugefile --encrypt-key "ALIAS1/BUCKET/issue=iliketobecrazybutnotsomuchreally" --insecure >/dev/null

$ mc cat ALIAS2/BUCKET/hugefile --encrypt-key "ALIAS2/BUCKET/issue=iliketobecrazybutnotsomuchreally" --insecure >/dev/null

$ mc cat ALIAS1/BUCKET/hugefile --encrypt-key "ALIAS1/BUCKET/issue=iliketobecrazybutnotsomuchreally" --insecure | wc -l
10000000

$ mc cat ALIAS2/BUCKET/hugefile --encrypt-key "ALIAS2/BUCKET/issue=iliketobecrazybutnotsomuchreally" --insecure | wc -l
10000000
```

8. Try the `hugefile` object to be loaded with different part size using command `mc put ./hugefile ALIAS1/BUCKET --encrypt-key "ALIAS1/BUCKET/hugefile=iliketobecrazybutnotsomuchreally" --insecure --part-size 50MiB` and replication of this object also should work as expected to other site

## Types of changes

## Checklist:
